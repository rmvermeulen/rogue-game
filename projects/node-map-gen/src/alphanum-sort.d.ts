declare module 'alphanum-sort' {
  // tslint:disable
  const sort: <T extends any[]>(input: T) => T;
  export = sort;
}
