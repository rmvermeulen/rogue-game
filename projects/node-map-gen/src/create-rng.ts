import * as Chance from 'chance';

export const createRNG = (seed?: number) => new Chance(seed);

export const rng = createRNG();
