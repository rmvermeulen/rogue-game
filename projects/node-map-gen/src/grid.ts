import * as sort from 'alphanum-sort';
import chalk from 'chalk';
import * as os from 'os';
import {
  addIndex,
  allPass,
  append,
  clone,
  defaultTo,
  equals,
  filter,
  flatten,
  identity,
  ifElse,
  is,
  join,
  last,
  lte,
  map,
  pipe,
  pluck,
  prepend,
  propEq,
  range,
  repeat,
  tap,
  times,
  trim,
  tryCatch,
  uniq,
  when,
  where,
  whereEq,
} from 'ramda';
import { random } from './random';
import { generate } from './utils';

const colors = (() => {
  const src = [
    chalk.red,
    chalk.redBright,
    chalk.green,
    chalk.greenBright,
    chalk.blue,
    chalk.blueBright,
    chalk.yellow,
    chalk.yellowBright,
    chalk.magenta,
    chalk.magentaBright,
    chalk.cyan,
    chalk.cyanBright,
  ];

  return random.shuffle(src);
})();

export interface IGridOptions extends Partial<Grid> {
  width: number;
  height: number;
  rooms: number;
  cells?: ICell[];
}

export interface ICell {
  id: number;
  x: number;
  y: number;
  room: number;
}

export interface IRoom {
  id: number;
  size: number;
  cells: ICell['id'][];
}

// represents a grid of cells
export class Grid {
  public readonly width: number;
  public readonly height: number;
  public readonly cells: ICell[];
  private readonly rooms: number;
  private constructor(options: IGridOptions) {
    const { width, height, rooms, cells } = options;
    const isClone = options instanceof Grid;
    console.assert(
      isClone === Boolean(cells),
      'Invalid Grid.CLONE: incomplete source',
    );

    this.width = width;
    this.height = height;
    this.rooms = rooms;
    this.cells = isClone ? clone(cells) : this.generateCells();
  }
  public static CREATE(options: IGridOptions) {
    return new Grid({ ...options });
  }

  public static CLONE(grid: Grid) {
    // tslint:disable-next-line: no-any
    return new Grid(grid as any);
  }
  public clone() {
    return Grid.CLONE(this);
  }

  public cellAt(x: number, y: number): undefined | ICell {
    // tslint:disable-next-line: no-any no-unsafe-any
    return this.cells.find((whereEq as any)({ x, y }));
  }

  // find all cells belonging to the given row
  public row(y: number): ICell[] {
    // tslint:disable-next-line: no-any no-unsafe-any
    return this.cells.filter((whereEq as any)({ y }));
  }

  // find all cells belonging to the given column
  public column(x: number): ICell[] {
    // tslint:disable-next-line: no-any no-unsafe-any
    return this.cells.filter((whereEq as any)({ x }));
  }

  // list the unique room-ids in the grid
  public listRooms(): number[] {
    return pipe<ICell[], number[], number[], number[]>(
      pluck('room'),
      uniq,
      filter(lte(0)),
    )(this.cells);
  }

  // get information on a specific room
  public findRoomById(id: IRoom['id']): IRoom | undefined {
    const cells = this.cells.filter(propEq('room', id));

    return cells.length > 0
      ? {
          id,
          size: cells.length,
          cells: pluck('id', cells),
        }
      : undefined;
  }

  /** render the grid into a string
   * example output:
   * +----+----+---+----+-----+
   * | 10   10 | 2   2  | 200 |
   * +----+    +   +    +-----+
   * | 0  | 10 | 2   2  | 10  |
   * +    +    +   +    +-----+
   * | 0  | 10 | 2   2  | 0   |
   * +    +    +   +    +     +
   * | 0  | 10 | 2   2  | 0   |
   * +    +----+---+----+-----+
   * | 0    0    0 | 10   10  |
   * +--- +----+---+----+-----+
   *
   * room 0 size=9 cells=4,5,10,14,15,19,20,21,22
   * room 1 size=8 cells=0,1,6,9,11,16,23,24
   * room 2 size=8 cells=2,3,7,8,12,13,17,18
   */
  // tslint:disable-next-line: max-func-body-length
  public display(useColors: boolean = true): string {
    const header = repeat('-', this.width);
    const lineFragments: (string | string[] | string[][])[] = [header];
    const columnNumberWidth: number[] = generate(this.width, () => 1);

    for (let y = 0; y < this.height; y += 1) {
      const line = [];
      const separatorLine = [];
      for (let x = 0; x < this.width; x += 1) {
        const cell = this.cellAt(x, y);
        const east = this.cellAt(x + 1, y);
        const south = this.cellAt(x, y + 1);

        // update the current column-width to the biggest in the column
        const roomStr = String(cell.room);
        columnNumberWidth[x] = Math.max(columnNumberWidth[x], roomStr.length);

        let str = roomStr;
        if (useColors) {
          const colorize = colors[cell.room % colors.length];
          str = colorize(roomStr);
        }

        const isSameRoom = when(Boolean, propEq('room', cell.room));
        line.push([
          str,
          (roomStr.length as unknown) as string /*cheat*/,
          isSameRoom(east) ? ' ' : '|',
        ]);
        separatorLine.push(isSameRoom(south) ? ' ' : '-');
      }
      lineFragments.push(line, separatorLine);
    }

    const rooms = sort(this.listRooms())
      .map(id => this.findRoomById(id))
      .filter(Boolean);

    if (rooms.length > 0) {
      lineFragments.push(
        [], // empty line to separate
        ...rooms.map(
          ({ id, size, cells }) => `room ${id} size=${size} cells=${cells}`,
        ),
      );
    }

    const mapIndexed = addIndex(map);
    const defaultTo1 = defaultTo(1);

    let previousWasVerticalWall = true;

    return pipe<typeof lineFragments, string[], string>(
      map(
        pipe(
          ifElse(
            is(String),
            identity,
            pipe(
              ifElse(
                ([first]) => is(Array, first),
                pipe(
                  mapIndexed(
                    (
                      [roomIdStr, roomIdStrLength, maybeWall]: [
                        string,
                        number,
                        '|' | ' '
                      ],
                      column,
                    ) => {
                      const width = defaultTo1(columnNumberWidth[column]);
                      // cannot use String.padEnd because of color in text
                      const paddedId =
                        roomIdStr + ' '.repeat(width - roomIdStrLength);

                      return `${paddedId} ${maybeWall} `;
                    },
                  ),
                  (list: string[]) => prepend('| ', list),
                ),
                pipe(
                  mapIndexed((wallType: '-' | ' ', column) => {
                    const width = defaultTo1(columnNumberWidth[column]);

                    const isVerticalWall = wallType === '-';

                    const dot =
                      isVerticalWall || previousWasVerticalWall
                        ? '+'
                        : column === 0
                        ? '|'
                        : ' ';

                    previousWasVerticalWall =
                      column === this.width - 1 ? isVerticalWall : false;

                    return dot + wallType.repeat(width + 2);
                  }),
                  (lines: string[]) =>
                    lines.length === 0
                      ? lines
                      : // try to read the last character of the last fragment
                      tryCatch(
                          pipe<string[], string, string>(
                            last,
                            last,
                          ),
                          () => ' ',
                        )(lines) === '-'
                      ? append('+', lines)
                      : append('|', lines),
                ),
              ),
              join(''),
            ),
          ),
          trim,
        ),
      ),
      join(os.EOL),
    )(lineFragments);
  }

  private generateCells() {
    const { width, height, rooms } = this;
    const avgRoomSize = Math.max(0, Math.floor((width * height) / rooms) - 1);
    const roomIds = times(n => repeat(n, avgRoomSize), rooms);
    let remainder = width * height - rooms * avgRoomSize;
    while (remainder > 0) {
      const id = random.natural({ max: rooms - 1 });
      roomIds[id].push(id);
      remainder -= 1;
    }

    const pool: number[] = random.shuffle(flatten(roomIds));

    const popSpecific = (n: number): number | undefined => {
      const i = pool.findIndex(equals(n));
      if (i === -1) {
        return undefined;
      }
      const [popped] = pool.splice(i, 1);

      return popped;
    };

    return pipe(
      map((id: number) => ({
        id,
        x: id % width,
        y: Math.floor(id / width),
        room: -1,
      })),
      tap((cells: ICell[]) => {
        cells.forEach((cell: ICell) => {
          const trySame = random.weighted([true, false], [15, 1]);
          let room: number;
          if (trySame) {
            // has valid room, which still has entries
            const neighbors = cells
              .filter(
                where({
                  room: allPass([
                    lte(0),
                    (r: ICell['room']) => pool.includes(r),
                  ]),
                }),
              )
              // manhattan distance
              .filter(
                ({ x, y }: typeof cell) =>
                  Math.abs(cell.x - x) + Math.abs(cell.y - y) === 1,
              );
            for (const n of neighbors) {
              const popped = popSpecific(n.room);
              if (popped !== undefined) {
                room = popped;
                break;
              }
            }
          }
          if (room === undefined) {
            room = pool.pop();
          }
          // update cell in-place
          cell.room = room;
        });
      }),
    )(range(0, width * height));
  }
}
