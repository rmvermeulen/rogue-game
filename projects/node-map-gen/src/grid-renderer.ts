import { assert } from 'chai';
import * as os from 'os';
import {
  aperture,
  compose,
  defaultTo,
  length,
  lensPath,
  map,
  path,
  pluck,
  repeat,
  set,
  sort,
  subtract,
  test,
  toString,
  unnest,
  when,
  zip,
} from 'ramda';
import {
  manyColors,
  manyColorsLength,
  someColors,
  someColorsLength,
} from './color-lists';
import { Grid, ICell } from './grid';
import { generate } from './utils';

export const renderSimple = (grid: Grid, useColors: boolean = true): string => {
  if (grid.listRooms().length > manyColorsLength) {
    // tslint:disable-next-line: no-parameter-reassignment
    useColors = false;
  }
  const colors = useColors
    ? grid.cells.length <= someColorsLength
      ? someColors()
      : manyColors()
    : [];

  const columnWidths = repeat(1, grid.width);

  generate(grid.height, (rowIndex: number) => {
    const row = grid.row(rowIndex);
    const roomStrWidths = compose<ICell[], number[], number[]>(
      map(
        compose<number, string, number>(
          length,
          toString,
        ),
      ),
      pluck<ICell, 'roomId'>('roomId'),
    )(row);
    // update column
    roomStrWidths.forEach((width, column) => {
      columnWidths[column] = Math.max(width, columnWidths[column]);
    });
  });

  return generate<string>(grid.height, (rowIndex: number) =>
    grid
      .row(rowIndex)
      .map((cell: ICell, index) => {
        let roomIdStr = cell.roomId.toString();
        const roomIdStrCharLength = roomIdStr.length;
        if (useColors) {
          const colorize = colors[cell.roomId % colors.length];
          roomIdStr = colorize(roomIdStr);
        }
        const width = columnWidths[index];

        // cannot use String.padEnd because of color in text
        return ' '.repeat(width - roomIdStrCharLength) + roomIdStr;
      })
      .join(' '),
  ).join(os.EOL);
};

// const stringifyLine = (
//   values: number[],
// ): { text: string; columnWidths: number[] } => {
//   return compose<number[], string[], { text: string; columnWidths: number[] }>(
//     applySpec({
//       text: (ss: string[]) => `| ${ss.join(' | ')} |`,
//       columnWidths: pluck('length'),
//     }),
//     map(String),
//   )(values);
// };

/** render the grid into a string
 * example output:
 * +----+----+---+----+-----+
 * | 10   10 | 2   2  | 200 |
 * +----+    +   +    +-----+
 * | 0  | 10 | 2   2  | 10  |
 * +    +    +   +    +-----+
 * | 0  | 10 | 2   2  | 0   |
 * +    +    +   +    +     +
 * | 0  | 10 | 2   2  | 0   |
 * +    +----+---+----+-----+
 * | 0    0    0 | 10   10  |
 * +--- +----+---+----+-----+
 *
 * room 0 size=9 cells=4,5,10,14,15,19,20,21,22
 * room 1 size=8 cells=0,1,6,9,11,16,23,24
 * room 2 size=8 cells=2,3,7,8,12,13,17,18
 */
// tslint:disable-next-line: max-func-body-length
export const render = (grid: Grid, useColors: boolean = true): string => {
  type WallStr = '|' | '+' | '-' | ' ';

  const columnWidths: number[] = generate(grid.width, () => 1);

  const mapLinesA = generate(grid.height, y => {
    const row = grid.row(y);
    const roomIds = pluck('roomId', row);
    generate(grid.width, column => {
      const tokenWidth = String(roomIds[column]).length;
      columnWidths[column] = Math.max(columnWidths[column], tokenWidth);
    });
    const walls: WallStr[] = [];
    for (const [a, b] of aperture(2, roomIds)) {
      const nextWall: WallStr = a === b ? ' ' : '|';
      walls.push(nextWall);
    }
    const [first, ...rest] = roomIds;

    return ['|', first, ...unnest(zip(walls, rest)), '|'];
  });

  const columnWalls: WallStr[][] = repeat([], grid.width);
  generate(grid.width, x => {
    const column = grid.column(x);
    const walls = columnWalls[x];
    walls.push('-');
    for (const [a, b] of aperture(2, pluck('roomId', column))) {
      const nextWall: WallStr = a === b ? ' ' : '-';
      walls.push(nextWall);
    }
    walls.push('-');
  });

  const topLine: WallStr[] = mapLinesA[0].map((value: number | WallStr) =>
    typeof value === 'number' ? '-' : value === '|' ? '+' : '-',
  );
  const lastMapLine = mapLinesA[grid.height - 1];
  const bottomLine: WallStr[] = lastMapLine.map((value: number | WallStr) =>
    typeof value === 'number' ? '-' : value === '|' ? '+' : '-',
  );
  const mapLinesB = [
    topLine,
    ...unnest(
      [...aperture(2, mapLinesA), [lastMapLine, undefined]].map(([a, b]) => {
        if (b === undefined) {
          return [a, bottomLine];
        }
        const walls = generate(a.length, x => {
          const ax = a[x];
          const bx = b[x];

          if (typeof ax === 'number') {
            return ax === bx ? ' ' : '-';
          }

          return ax === bx ? '|' : '+';
        });

        return [a, walls];
      }),
    ),
  ];

  assert.lengthOf(mapLinesB, grid.height * 2 + 1);

  let updated = mapLinesB;

  generate(grid.height + 1, gy => {
    generate(grid.width + 1, gx => {
      const x = gx * 2;
      const y = gy * 2;
      // const cornerPart = mapLinesB[y * 2][x * 2];
      const [n, s, e, w] = [[y - 1, x], [y + 1, x], [y, x - 1], [y, x + 1]]
        // tslint:disable-next-line: no-any no-unsafe-any
        .map(p => path(p as any[], mapLinesB))
        .map(defaultTo(' '));
      const assignField = (char: string) => {
        // tslint:disable-next-line: no-any
        const modify = set(lensPath([y, x]), char) as any;
        // tslint:disable-next-line: no-unsafe-any
        updated = modify(updated);
      };
      if (n === ' ' && s === ' ' && e === ' ' && w === ' ') {
        return assignField(' ');
      }
      if (e === ' ' && w === ' ') {
        return assignField('|');
      }
      if (n === ' ' && s === ' ') {
        return assignField('-');
      }

      return assignField('+');
    });
  });

  const lineFragments = updated.map((chunks, y) => {
    const newLine = chunks.map(String).map((chunk, x) => {
      const isRoomCenterX = x % 2 === 1;
      if (!isRoomCenterX) {
        return chunk;
      }
      const isRoomCenterY = y % 2 === 1;
      const width = defaultTo(1, columnWidths[(x - 1) / 2]);

      const padded = isRoomCenterY
        ? ` ${chunk.padStart(width)} `
        : chunk.repeat(width + 2);
      assert.lengthOf(padded, width + 2);

      return padded;
    });

    return newLine.join('');
  });

  const rooms = sort(subtract, grid.listRooms())
    .map(id => grid.findRoomById(id))
    .filter(Boolean);

  const roomFragments = rooms.map(
    ({ id, size, cells }) => `room ${id} size=${size} cells=${cells}`,
  );

  const renderedString = lineFragments.join(os.EOL);

  if (useColors) {
    const colors =
      grid.cells.length <= someColorsLength ? someColors() : manyColors();
    const getColorizer = (n: number) => colors[n % colors.length];

    return [
      renderedString
        .split(' ')
        .map(
          when(test(/^\d+$/), (str: string) => getColorizer(Number(str))(str)),
        )
        .join(' '),
      '',
      ...roomFragments,
    ].join(os.EOL);
  }

  return [renderedString, '', ...roomFragments].join(os.EOL);
};
