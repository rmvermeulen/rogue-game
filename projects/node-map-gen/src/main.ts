import { Grid } from './grid';

export const gridFactory = Grid.create;
