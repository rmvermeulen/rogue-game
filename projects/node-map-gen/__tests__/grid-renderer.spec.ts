// tslint:disable no-import-side-effect no-implicit-dependencies
import 'jest-extended';
import '../src/strip-ansi.d';
// tslint:enable no-import-side-effect no-implicit-dependencies

import * as os from 'os';
import {
  compose,
  converge,
  length,
  split,
  takeLastWhile,
  takeWhile,
  test as testRE,
  uniqBy,
} from 'ramda';
// tslint:disable-next-line: no-implicit-dependencies
import * as stripAnsi from 'strip-ansi';
import { Grid } from '../src/grid';
import { render, renderSimple } from '../src/grid-renderer';

const toLines: (str: string) => string[] = split(os.EOL);

// const countSubstr = (str: string) =>
//   compose(
//     x => x - 1,
//     length,
//     split(str),
//   );

describe.each(
  [
    [5, 5, 4],
    [10, 5, 4],
    [5, 10, 4],
    [10, 10, 4],
    [5, 5, 16],
    [10, 5, 16],
    [5, 10, 16],
    [10, 10, 16],
  ].slice(0, 1),
)('Grid renderer', (width, height, roomCount) => {
  let grid: Grid;
  beforeAll(() => {
    grid = Grid.CREATE({
      width,
      height,
      roomCount,
    });
    expect(grid).toBeDefined();
  });

  test('simple render', () => {
    const display = renderSimple(grid);
    expect(display).toBeString();
    expect(display).toMatchSnapshot();
    expect(stripAnsi(display)).toMatchSnapshot();
    // all lines have the same length
    expect(
      compose(
        uniqBy(length),
        toLines,
        stripAnsi as (s: string) => string,
      )(display),
    ).toHaveLength(1);
  });

  test('fancy render', () => {
    const display = render(grid);
    expect(display).toBeString();
    // expect(display).toMatchSnapshot();
    // expect(stripAnsi(display)).toMatchSnapshot();

    // display format
    // map line 1
    // map line 2
    // <empty line>
    // room line 2
    // room line 2

    const [mapLines, roomLines] = compose<
      string,
      string,
      string[],
      [string[], string[]]
    >(
      converge((a, b) => [a, b], [
        takeWhile<string>(Boolean),
        takeLastWhile<string>(Boolean),
      ]),
      toLines,
      stripAnsi,
    )(display); // split on empty line

    // all map-lines have the same length
    expect(uniqBy(length, mapLines)).toHaveLength(1);

    // const [rooms, walls] = partition(testRE(/\d/), mapLines);

    // const hWalls = rooms.map(countSubstr('|'));
    // expect(hWalls).not.toSatisfyAll(equals(width + 1));

    // const vWalls = walls.map(countSubstr('+'));
    // expect(hWalls).not.toSatisfyAll(equals(height + 1));

    // room line regexp check
    expect(roomLines).toSatisfyAll(testRE(/room \d+ size=\d+ cells=(\d+\,?)+/));
  });
});
