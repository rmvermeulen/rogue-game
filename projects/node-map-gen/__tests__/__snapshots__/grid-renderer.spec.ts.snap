// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renderer simple render 1`] = `
"[93m0[39m [93m0[39m [102m[30m3[39m[49m [93m0[39m [96m2[39m
[93m0[39m [102m[30m3[39m[49m [94m1[39m [94m1[39m [102m[30m3[39m[49m
[93m0[39m [96m2[39m [102m[30m3[39m[49m [96m2[39m [96m2[39m
[94m1[39m [94m1[39m [96m2[39m [102m[30m3[39m[49m [102m[30m3[39m[49m
[94m1[39m [94m1[39m [102m[30m3[39m[49m [93m0[39m [93m0[39m"
`;

exports[`Grid renderer simple render 2`] = `
"0 0 3 0 2
0 3 1 1 3
0 2 3 2 2
1 1 2 3 3
1 1 3 0 0"
`;

exports[`Grid renderer simple render 3`] = `
"[102m[30m3[39m[49m [95m1[39m [102m[30m3[39m[49m [92m0[39m [92m0[39m [95m1[39m [103m[30m2[39m[49m [92m0[39m [103m[30m2[39m[49m [95m1[39m
[103m[30m2[39m[49m [92m0[39m [103m[30m2[39m[49m [102m[30m3[39m[49m [95m1[39m [95m1[39m [103m[30m2[39m[49m [95m1[39m [102m[30m3[39m[49m [92m0[39m
[102m[30m3[39m[49m [102m[30m3[39m[49m [102m[30m3[39m[49m [92m0[39m [103m[30m2[39m[49m [95m1[39m [92m0[39m [103m[30m2[39m[49m [102m[30m3[39m[49m [95m1[39m
[103m[30m2[39m[49m [92m0[39m [92m0[39m [102m[30m3[39m[49m [102m[30m3[39m[49m [92m0[39m [103m[30m2[39m[49m [103m[30m2[39m[49m [95m1[39m [102m[30m3[39m[49m
[103m[30m2[39m[49m [95m1[39m [103m[30m2[39m[49m [92m0[39m [95m1[39m [92m0[39m [95m1[39m [92m0[39m [103m[30m2[39m[49m [102m[30m3[39m[49m"
`;

exports[`Grid renderer simple render 4`] = `
"3 1 3 0 0 1 2 0 2 1
2 0 2 3 1 1 2 1 3 0
3 3 3 0 2 1 0 2 3 1
2 0 0 3 3 0 2 2 1 3
2 1 2 0 1 0 1 0 2 3"
`;

exports[`Grid renderer simple render 5`] = `
"[46m[30m0[39m[49m [104m[30m2[39m[49m [105m[30m1[39m[49m [104m[30m2[39m[49m [104m[30m2[39m[49m
[105m[30m1[39m[49m [104m[30m2[39m[49m [105m[30m1[39m[49m [95m3[39m [105m[30m1[39m[49m
[105m[30m1[39m[49m [95m3[39m [95m3[39m [104m[30m2[39m[49m [46m[30m0[39m[49m
[104m[30m2[39m[49m [104m[30m2[39m[49m [104m[30m2[39m[49m [95m3[39m [95m3[39m
[46m[30m0[39m[49m [104m[30m2[39m[49m [95m3[39m [105m[30m1[39m[49m [95m3[39m
[105m[30m1[39m[49m [105m[30m1[39m[49m [104m[30m2[39m[49m [105m[30m1[39m[49m [46m[30m0[39m[49m
[46m[30m0[39m[49m [46m[30m0[39m[49m [95m3[39m [95m3[39m [46m[30m0[39m[49m
[46m[30m0[39m[49m [104m[30m2[39m[49m [46m[30m0[39m[49m [46m[30m0[39m[49m [46m[30m0[39m[49m
[104m[30m2[39m[49m [95m3[39m [95m3[39m [95m3[39m [105m[30m1[39m[49m
[105m[30m1[39m[49m [95m3[39m [95m3[39m [105m[30m1[39m[49m [46m[30m0[39m[49m"
`;

exports[`Grid renderer simple render 6`] = `
"0 2 1 2 2
1 2 1 3 1
1 3 3 2 0
2 2 2 3 3
0 2 3 1 3
1 1 2 1 0
0 0 3 3 0
0 2 0 0 0
2 3 3 3 1
1 3 3 1 0"
`;

exports[`Grid renderer simple render 7`] = `
"[106m[30m2[39m[49m [106m[30m2[39m[49m [106m[30m2[39m[49m [31m3[39m [44m[30m1[39m[49m [31m3[39m [36m0[39m [31m3[39m [31m3[39m [36m0[39m
[36m0[39m [106m[30m2[39m[49m [36m0[39m [36m0[39m [36m0[39m [44m[30m1[39m[49m [106m[30m2[39m[49m [106m[30m2[39m[49m [36m0[39m [31m3[39m
[44m[30m1[39m[49m [31m3[39m [36m0[39m [106m[30m2[39m[49m [106m[30m2[39m[49m [36m0[39m [31m3[39m [36m0[39m [106m[30m2[39m[49m [44m[30m1[39m[49m
[44m[30m1[39m[49m [36m0[39m [31m3[39m [106m[30m2[39m[49m [31m3[39m [44m[30m1[39m[49m [36m0[39m [106m[30m2[39m[49m [36m0[39m [36m0[39m
[31m3[39m [44m[30m1[39m[49m [36m0[39m [106m[30m2[39m[49m [31m3[39m [44m[30m1[39m[49m [31m3[39m [36m0[39m [106m[30m2[39m[49m [44m[30m1[39m[49m
[31m3[39m [31m3[39m [44m[30m1[39m[49m [36m0[39m [44m[30m1[39m[49m [44m[30m1[39m[49m [31m3[39m [44m[30m1[39m[49m [106m[30m2[39m[49m [44m[30m1[39m[49m
[106m[30m2[39m[49m [31m3[39m [44m[30m1[39m[49m [36m0[39m [44m[30m1[39m[49m [31m3[39m [106m[30m2[39m[49m [36m0[39m [36m0[39m [44m[30m1[39m[49m
[31m3[39m [44m[30m1[39m[49m [44m[30m1[39m[49m [36m0[39m [106m[30m2[39m[49m [106m[30m2[39m[49m [31m3[39m [106m[30m2[39m[49m [31m3[39m [106m[30m2[39m[49m
[36m0[39m [36m0[39m [44m[30m1[39m[49m [36m0[39m [36m0[39m [36m0[39m [31m3[39m [106m[30m2[39m[49m [31m3[39m [31m3[39m
[106m[30m2[39m[49m [44m[30m1[39m[49m [106m[30m2[39m[49m [44m[30m1[39m[49m [36m0[39m [106m[30m2[39m[49m [44m[30m1[39m[49m [44m[30m1[39m[49m [31m3[39m [31m3[39m"
`;

exports[`Grid renderer simple render 8`] = `
"2 2 2 3 1 3 0 3 3 0
0 2 0 0 0 1 2 2 0 3
1 3 0 2 2 0 3 0 2 1
1 0 3 2 3 1 0 2 0 0
3 1 0 2 3 1 3 0 2 1
3 3 1 0 1 1 3 1 2 1
2 3 1 0 1 3 2 0 0 1
3 1 1 0 2 2 3 2 3 2
0 0 1 0 0 0 3 2 3 3
2 1 2 1 0 2 1 1 3 3"
`;

exports[`Grid renderer simple render 9`] = `
"[103m[30m6[39m[49m [96m10[39m [32m12[39m [35m0[39m [103m[30m6[39m[49m
[104m[30m2[39m[49m [96m10[39m  [94m3[39m [33m7[39m [46m[30m4[39m[49m
[94m3[39m  [94m3[39m  [94m3[39m [106m[30m5[39m[49m [36m8[39m
[104m[30m2[39m[49m [96m10[39m  [36m8[39m [46m[30m4[39m[49m [35m0[39m
[46m[30m4[39m[49m  [35m0[39m  [46m[30m4[39m[49m [104m[30m2[39m[49m [46m[30m4[39m[49m"
`;

exports[`Grid renderer simple render 10`] = `
"6 10 12 0 6
2 10  3 7 4
3  3  3 5 8
2 10  8 4 0
4  0  4 2 4"
`;

exports[`Grid renderer simple render 11`] = `
"[96m1[39m  [106m[30m7[39m[49m [91m10[39m [33m13[39m [92m12[39m  [95m6[39m [92m12[39m [33m13[39m [91m10[39m  [96m1[39m
[35m8[39m  [36m4[39m  [95m6[39m  [42m[30m0[39m[49m [33m13[39m  [106m[30m7[39m[49m  [95m6[39m [92m12[39m  [35m8[39m  [93m2[39m
[43m[30m5[39m[49m [92m12[39m [46m[30m11[39m[49m  [95m6[39m [33m13[39m  [43m[30m5[39m[49m  [45m[30m9[39m[49m  [42m[30m0[39m[49m [46m[30m11[39m[49m  [36m4[39m
[42m[30m0[39m[49m [32m14[39m  [43m[30m5[39m[49m  [96m1[39m [46m[30m11[39m[49m [32m14[39m  [44m[30m3[39m[49m [91m10[39m  [44m[30m3[39m[49m  [106m[30m7[39m[49m
[35m8[39m  [45m[30m9[39m[49m  [44m[30m3[39m[49m [46m[30m11[39m[49m [34m15[39m  [93m2[39m  [45m[30m9[39m[49m [33m13[39m  [95m6[39m [34m15[39m"
`;

exports[`Grid renderer simple render 12`] = `
"1  7 10 13 12  6 12 13 10  1
8  4  6  0 13  7  6 12  8  2
5 12 11  6 13  5  9  0 11  4
0 14  5  1 11 14  3 10  3  7
8  9  3 11 15  2  9 13  6 15"
`;

exports[`Grid renderer simple render 13`] = `
"[95m10[39m [102m[30m11[39m[49m  [45m[30m8[39m[49m  [31m1[39m  [91m5[39m
[34m15[39m  [36m9[39m [33m14[39m  [105m[30m4[39m[49m  [46m[30m6[39m[49m
[34m15[39m  [35m2[39m [33m14[39m  [44m[30m7[39m[49m  [105m[30m4[39m[49m
 [36m9[39m [92m13[39m [106m[30m12[39m[49m  [105m[30m4[39m[49m  [35m2[39m
 [45m[30m8[39m[49m [95m10[39m [106m[30m12[39m[49m [95m10[39m  [35m2[39m
 [101m[30m3[39m[49m [33m14[39m [92m13[39m [92m13[39m  [93m0[39m
 [91m5[39m [102m[30m11[39m[49m  [46m[30m6[39m[49m  [101m[30m3[39m[49m  [101m[30m3[39m[49m
 [31m1[39m  [105m[30m4[39m[49m  [101m[30m3[39m[49m  [105m[30m4[39m[49m  [105m[30m4[39m[49m
[106m[30m12[39m[49m  [44m[30m7[39m[49m  [101m[30m3[39m[49m  [93m0[39m [92m13[39m
 [105m[30m4[39m[49m [106m[30m12[39m[49m  [44m[30m7[39m[49m  [91m5[39m [33m14[39m"
`;

exports[`Grid renderer simple render 14`] = `
"10 11  8  1  5
15  9 14  4  6
15  2 14  7  4
 9 13 12  4  2
 8 10 12 10  2
 3 14 13 13  0
 5 11  6  3  3
 1  4  3  4  4
12  7  3  0 13
 4 12  7  5 14"
`;

exports[`Grid renderer simple render 15`] = `
"[96m13[39m [96m13[39m  [41m[30m0[39m[49m [96m13[39m  [92m8[39m [101m[30m15[39m[49m  [103m[30m1[39m[49m [101m[30m15[39m[49m [33m14[39m [95m11[39m
[95m11[39m  [34m3[39m  [34m3[39m  [32m6[39m [96m13[39m  [42m[30m7[39m[49m [95m11[39m [101m[30m15[39m[49m  [31m4[39m [35m12[39m
[101m[30m15[39m[49m  [32m6[39m  [41m[30m0[39m[49m  [31m4[39m  [103m[30m1[39m[49m  [42m[30m7[39m[49m  [31m4[39m  [93m2[39m [95m11[39m [33m14[39m
 [42m[30m7[39m[49m [95m11[39m  [41m[30m0[39m[49m [96m13[39m  [34m3[39m  [41m[30m0[39m[49m  [34m3[39m  [31m4[39m  [104m[30m5[39m[49m  [103m[30m1[39m[49m
 [42m[30m7[39m[49m  [104m[30m5[39m[49m  [43m[30m9[39m[49m [33m14[39m  [104m[30m5[39m[49m  [34m3[39m  [34m3[39m [106m[30m10[39m[49m  [93m2[39m  [43m[30m9[39m[49m
 [41m[30m0[39m[49m  [104m[30m5[39m[49m  [43m[30m9[39m[49m [35m12[39m  [41m[30m0[39m[49m  [103m[30m1[39m[49m  [93m2[39m  [32m6[39m [96m13[39m [101m[30m15[39m[49m
 [92m8[39m [106m[30m10[39m[49m [106m[30m10[39m[49m  [92m8[39m  [92m8[39m [35m12[39m [33m14[39m  [93m2[39m [33m14[39m  [93m2[39m
 [92m8[39m  [93m2[39m  [92m8[39m  [104m[30m5[39m[49m [96m13[39m  [32m6[39m  [32m6[39m  [32m6[39m [106m[30m10[39m[49m  [93m2[39m
[95m11[39m  [104m[30m5[39m[49m  [34m3[39m  [41m[30m0[39m[49m [35m12[39m  [103m[30m1[39m[49m [95m11[39m [33m14[39m  [43m[30m9[39m[49m  [43m[30m9[39m[49m
[35m12[39m  [31m4[39m [35m12[39m  [103m[30m1[39m[49m  [31m4[39m  [43m[30m9[39m[49m [35m12[39m [101m[30m15[39m[49m [106m[30m10[39m[49m  [42m[30m7[39m[49m"
`;

exports[`Grid renderer simple render 16`] = `
"13 13  0 13  8 15  1 15 14 11
11  3  3  6 13  7 11 15  4 12
15  6  0  4  1  7  4  2 11 14
 7 11  0 13  3  0  3  4  5  1
 7  5  9 14  5  3  3 10  2  9
 0  5  9 12  0  1  2  6 13 15
 8 10 10  8  8 12 14  2 14  2
 8  2  8  5 13  6  6  6 10  2
11  5  3  0 12  1 11 14  9  9
12  4 12  1  4  9 12 15 10  7"
`;
